addons/product_margin/models/product_product.py

```diff
@@ -128,38 +128,43 @@ def _compute_product_margin_fields_values(self, field_names=None):
-            result = self.env.cr.fetchall()[0]
-            res[val.id]['purchase_avg_price'] = result[0] and result[0] or 0.0
-            res[val.id]['purchase_num_invoiced'] = result[1] and result[1] or 0.0
-            res[val.id]['total_cost'] = result[2] and result[2] or 0.0
-            res[val.id]['normal_cost'] = val.standard_price * res[val.id]['purchase_num_invoiced']
-            res[val.id]['purchase_gap'] = res[val.id]['normal_cost'] - res[val.id]['total_cost']
+        invoice_types = ('out_invoice', 'out_refund')
+        self.env.cr.execute(sqlstr, (tuple(self.ids), states, invoice_payment_states, invoice_types, date_from, date_to, company_id))
+        for product_id, avg, qty, total, sale in self.env.cr.fetchall():
+            res[product_id]['sale_avg_price'] = avg and avg or 0.0
+            res[product_id]['sale_num_invoiced'] = qty and qty or 0.0
+            res[product_id]['turnover'] = total and total or 0.0
+            res[product_id]['sale_expected'] = sale and sale or 0.0
+            res[product_id]['sales_gap'] = res[product_id]['sale_expected'] - res[product_id]['turnover']
+            res[product_id]['total_margin'] = res[product_id]['turnover']
+            res[product_id]['expected_margin'] = res[product_id]['sale_expected']
+            res[product_id]['total_margin_rate'] = res[product_id]['turnover'] and res[product_id]['total_margin'] * 100 / res[product_id]['turnover'] or 0.0
+            res[product_id]['expected_margin_rate'] = res[product_id]['sale_expected'] and res[product_id]['expected_margin'] * 100 / res[product_id]['sale_expected'] or 0.0
 
-            res[val.id]['total_margin'] = res[val.id]['turnover'] - res[val.id]['total_cost']
-            res[val.id]['expected_margin'] = res[val.id]['sale_expected'] - res[val.id]['normal_cost']
-            res[val.id]['total_margin_rate'] = res[val.id]['turnover'] and res[val.id]['total_margin'] * 100 / res[val.id]['turnover'] or 0.0
-            res[val.id]['expected_margin_rate'] = res[val.id]['sale_expected'] and res[val.id]['expected_margin'] * 100 / res[val.id]['sale_expected'] or 0.0
-            for k, v in res[val.id].items():
-                setattr(val, k, v)
+        ctx = self.env.context.copy()
+        ctx['force_company'] = company_id
+        invoice_types = ('in_invoice', 'in_refund')
+        self.env.cr.execute(sqlstr, (tuple(self.ids), states, invoice_payment_states, invoice_types, date_from, date_to, company_id))
+        for product_id, avg, qty, total, dummy in self.env.cr.fetchall():
+            res[product_id]['purchase_avg_price'] = avg and avg or 0.0
+            res[product_id]['purchase_num_invoiced'] = qty and qty or 0.0
+            res[product_id]['total_cost'] = total and total or 0.0
+            res[product_id]['normal_cost'] = self.browse(product_id).standard_price * res[product_id]['purchase_num_invoiced']
```

Comment:
that browse() in a loop is probably hurting the performances you try to improve. Did you benchmark it again after the changes you made?

Reply:
Yeah you're right. I moved some of the computations to the `for product in self` loop below to avoid calling `self.browse` here.